feat(reminders): request notification permissions

ericli3690 opened the pull request:
**Based on top of, and thus blocked by, the following PRs:**
- https://github.com/ankidroid/Anki-Android/pull/18978
- https://github.com/ankidroid/Anki-Android/pull/19109/

## Purpose / Description
Creates a BottomSheet for requesting notification permissions from the user. This BottomSheet shows up when the user first triggers a sync and when the user first creates a review reminder, as these are the two features that rely on notifications.

**Changes from all commits:**
As discussed in Discord, we have decided to use the PermissionsActivity which shows up when the app is initially installed to request mandatory permissions from the user. For optional permissions, we have decided to try and request them in a less intrusive way. Here, I've created a PermissionsBottomSheet that can pop up and request optional permissions from the user. I use this bottom sheet in a later commit for requesting notification permissions.

I've modeled PermissionsBottomSheet after PermissionsActivity. I'm fairly sure I cannot reuse PermissionsActivity's code because it is a full-screen activity, whereas PermissionsBottomSheet needs to inherit BottomSheetFragment and be a fragment.

Created the PermissionsFragment which will be hosted inside PermissionsBottomSheet for requesting notification permissions from the user. It is launched via `requestNotificationsPermissionIfNeeded` in Permissions. This method takes a callback so that I can do some conditional preference-setting in a later commit. Created a new PermissionSet for the notifications permission in InitialActivity, which is gated behind API 33. In a way, this means I've sort of resurrected TiramisuPermissionsFragment, so maybe I shouldn't have been so quick to delete it, whoops.

Adds two new boolean Preferences for saving whether we have requested notification permissions from the user: one for whether it has been done for the sync process yet (we will request notification permissions the first time the user tries to sync) and one for whether it has been done for the review reminders feature yet (we will request notification permissions the first time the user creates a review reminder).

I'm aware there is a `shouldShowRequestPermissionRationale` method which Google recommends using to check if we should request notification permissions from the user. I do not believe we should use it as it is incredibly confusing and has some very weird edge-case behaviour. See [this post](https://stackoverflow.com/questions/41310510/what-is-the-difference-between-shouldshowrequestpermissionrationale-and-requestp) for more information. Rather, it makes good sense to use two quick preferences and carefully handle whether we have shown the request dialog ourselves. This is also more extensible in the future, should a future developer choose to request the notification permission in a different circumstance, too.

I have decided to request the permission once for the sync feature and once for the review reminder feature because if the user denies the permission for the sync feature and then goes to enable review reminders, they may change their mind about providing AnkiDroid with notification permissions; and vice versa.

If the notification is denied, I have chosen not to downgrade any review reminders functionality: the user will still be able to create, edit, and delete them. However, they will not receive any notifications. I do not believe it makes sense to restrict the user's ability to edit review reminders even though they have notifications disabled, as a user might want to customize their reminders to their liking before enabling notifications, etc. In the future, I may add a help screen to help users who are confused why they are not getting notifications and who may have notifications accidentally disabled, but otherwise, I think we should assume that if the user has denied notification permissions, then we simply should not sent notifications, and do not need to downgrade any other functionality.

The old code requested notification permissions via MyAccount and requests the permission each time a new user logs in, which makes no sense. Notifications are tied to the phone the app is installed on, not an individual user profile. Rather, we should be requesting notification permissions when the user first triggers a sync. This is what this change accomplishes.

Deletes all old logic for requesting notification permissions. Deletes `checkNotificationPermission` as it has been moved to Permissions as `shouldRequestNotificationPermissions`. Requests notification permissions the first time the user triggers a sync.

Request notification permissions when the user first creates a review reminder. Involves a simple edit to AddEditReminderDialog.

## Fixes
* Fixes https://github.com/ankidroid/Anki-Android/issues/15952
* Part of GSoC 2025: Review Reminders

## Approach
- Creates a new PermissionsBottomSheet. My intention is: mandatory permissions -> use PermissionsActivity; optional permissions -> use PermissionsBottomSheet. I think the BottomSheet is less intrusive compared to the full-screen PermissionsActivity, as befits an optional permission.

## How Has This Been Tested?
- Runs on a physical Samsung S23, API 34.
- Medium Phone, API 35.

## Learning (optional, can help others)
- Spent a lot of time stuck on a bug involving fragmentManagers, but managed to figure it out.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

<!--- Uncomment this section ONLY if this PR introduces new resources (external libraries, icons etc)
## Licenses
_For each new external resource, add a row to the table below:_

| Library | Description | License |
| --- | --- | --- |
| Sample Icon Library | Sample Description | [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt) |

**Maintainers:**

* [ ] Add the https://github.com/ankidroid/Anki-Android/labels/Licenses label
* [ ] Update the [licenses](https://github.com/ankidroid/Anki-Android/wiki/Licences) wiki when merging
--->

ericli3690 commented:
- Fixed the headline of the BottomSheet, which was hardcoded for some reason. It now reads: `AnkiDroid works best with these permissions`.
- Rebased on top of https://github.com/ankidroid/Anki-Android/pull/18978

ericli3690 commented:
- Rebased.

ericli3690 commented:
- Rebased. The optional permissions header has been moved to https://github.com/ankidroid/Anki-Android/pull/18978, which this PR is based on top of.

ericli3690 commented:
- Rebased and resolved merge conflicts.
- Now only blocked by https://github.com/ankidroid/Anki-Android/pull/19109.

ericli3690 commented:
Dependencies have been merged. Rebased and ready for review!

ericli3690 commented:
Ashish has requested that I move the close button to the top right corner, which makes sense to me. I'll also look into potentially turning it into an "X" icon.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -2121,7 +2117,12 @@ open class DeckPicker :
 
-        MyAccount.checkNotificationPermission(this, notificationPermissionLauncher)
+        // Request notification permissions from the user if they have not been requested ever before
+        if (!Prefs.syncNotifsRequestShown) {
```
@david-allison Yep, I've thought about that before. From the PR description:

> I'm aware there is a shouldShowRequestPermissionRationale method which Google recommends using to check if we should request notification permissions from the user. I do not believe we should use it as it is incredibly confusing and has some very weird edge-case behaviour. See [this post](https://stackoverflow.com/questions/41310510/what-is-the-difference-between-shouldshowrequestpermissionrationale-and-requestp) for more information. Rather, it makes good sense to use two quick preferences and carefully handle whether we have shown the request dialog ourselves. This is also more extensible in the future, should a future developer choose to request the notification permission in a different circumstance, too.

Basically, the internal workings of `shouldShowRequestPermissionRationale` are kind of counterintuitive and the name isn't really self-documenting. Google really should have designed it to return an enum instead of a boolean. See [the top answer here](https://stackoverflow.com/questions/41310510/what-is-the-difference-between-shouldshowrequestpermissionrationale-and-requestp) and [the top answer here](https://stackoverflow.com/questions/56412401/working-of-shouldshowrequestpermissionrationale-method) for more information. The function only returns true if the app has requested the permission before and the user has clicked "deny" without clicking "don't ask again". It returns false when it's requesting a permission for the first time. I'd like to show the AnkiDroid notification request when the user performs an action that requires the permission for the first time, so `shouldShowRequestPermissionRationale` doesn't work for this use case. That's why I created a Pref.

I've also considered adding another case to trigger the permission request BottomSheet if the user has clicked "deny" without clicking "don't ask again", which we could use `shouldShowRequestPermissionRationale` for (i.e. we would show the BottomSheet both if the user is opening the app for the first time, via the Pref, OR if the user has clicked "deny" without clicking "don't ask again", via `shouldShowRequestPermissionRationale`). The problem with that is I don't see it happening very often. With this PR, the only times we request notification permissions from the user, we do it through the BottomSheet above. For a user to be logged as having clicked "deny" without clicking "don't ask again", they would have to 1. trigger the BottomSheet, 2. read the message, 3. click on the enabling toggle, and then 4. click deny. I don't see many users changing their minds after clicking the toggle to grant notification permissions (which is why, in that scenario, the BottomSheet currently opens up the OS settings to allow the user to grant notification permissions directly, assuming they performed a misclick or that something went wrong with launching the OS permission request dialog).

I will concede that the fact that we only show the BottomSheet at most two times to users (once on the first sync and once on the first review reminder created) is a bit weird. Perhaps we should have the BottomSheet have a chance to show up again if a user who hasn't granted notification permissions performs an action that requires notification permissions. For example, we currently only request notification permissions at most once when the user first performs a sync. Maybe we can request it again on the 5th sync and the 10th sync, and only then fully give up? This has the potential to be annoying to the user, though. Similarly with requesting notification permissions when the user creates review reminders: we currently only request permissions when the user creates a review reminder for the first time, and can potentially change it so that it requests it again on the 5th and 10th review reminder created, and never again after that. This change would mean editing the `Prefs.syncNotifsRequestShown` to be an Int instead of a Boolean, and incrementing it based on how many times the permission request BottomSheet has been shown.

Thoughts?

ericli3690 commented:
Rebased and resolved merge conflicts!

@criticalAY I noticed you made major changes to the login screens in your [recently-merged PR](https://github.com/ankidroid/Anki-Android/pull/17908). I've resolved my merge conflicts and re-implemented my feature by deleting some of your code. Please let me know if I've accidentally modified anything important that you've written -- I think I was pretty careful, though!

Also, @mikehardy, any thoughts on `shouldShowRequestPermissionRationale` over [here](https://github.com/ankidroid/Anki-Android/pull/19167#discussion_r2465683635)?

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsBottomSheet.kt:
```diff
@@ -0,0 +1,100 @@
+        savedInstanceState: Bundle?,
+    ) {
+        super.onViewCreated(view, savedInstanceState)
+        val closeButton = view.findViewById<ImageButton>(R.id.close_button)
```
Since we already have viewBinding we can switch to the same, @david-allison your opinion?

```PATCH
Subject: [PATCH] Fragment ViewBinding
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsBottomSheet.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsBottomSheet.kt b/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsBottomSheet.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsBottomSheet.kt	(revision 6599421ad87aa12e015f76afd7e729628f1b0483)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsBottomSheet.kt	(date 1762164037747)
@@ -29,6 +29,8 @@
 import com.google.android.material.bottomsheet.BottomSheetDialogFragment
 import com.ichi2.anki.PermissionSet
 import com.ichi2.anki.R
+import com.ichi2.anki.databinding.PermissionsBottomSheetBinding
+import dev.androidbroadcast.vbpd.viewBinding
 
 /**
  * BottomSheet that requests permissions from the user.
@@ -40,6 +42,8 @@
  */
 @RequiresApi(Build.VERSION_CODES.TIRAMISU)
 class PermissionsBottomSheet : BottomSheetDialogFragment() {
+    private val binding by viewBinding(PermissionsBottomSheetBinding::bind)
+
     override fun onCreateView(
         inflater: LayoutInflater,
         container: ViewGroup?,
@@ -51,8 +55,8 @@
         savedInstanceState: Bundle?,
     ) {
         super.onViewCreated(view, savedInstanceState)
-        val closeButton = view.findViewById<ImageButton>(R.id.close_button)
-        closeButton.setOnClickListener { dismiss() }
+
+        binding.closeButton.setOnClickListener { dismiss() }
 
         val permissionSet =
             requireNotNull(BundleCompat.getParcelable(requireArguments(), PERMISSION_SET_ARGUMENT_KEY, PermissionSet::class.java)) {
@@ -62,10 +66,9 @@
             requireNotNull(permissionSet.permissionsFragment?.getDeclaredConstructor()?.newInstance()) {
                 "invalid permissionsFragment"
             }
-        view.post {
-            childFragmentManager.commit {
-                replace(R.id.bottom_sheet_fragment_container, permissionsFragment)
-            }
+
+        childFragmentManager.commit {
+            replace(binding.bottomSheetFragmentContainer.id, permissionsFragment)
         }
     }
 
```

david-allison replied:
Implementers choice, I'd support it strongly, but ViewBindings should NOT be blocking/slowing down PRs unless the implementer is onboard with it.

I'm taking responsibility for migrating the app to use ViewBindings. It's ultimately my responsibility to take this on in a follow-up PR

* #11116

mikehardy reviewed (commented):
pretty sweet

- one long thought about the impact of app data backup / restore on to new devices, possibly resolve-able by noting it and ignoring it (that's what I'd do, on balance - but if and only if we have a good UX for "you're editing reminders but notifications are disabled"
- one thought on how to handle the permissions request just a touch better, to avoid asking the system to pop a dialog when it's invalid

ericli3690 commented:
On me -- will try to respond and revise this PR this weekend, but no promises as I'm entering final exam season where I'm at. Thanks!!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/utils/Permissions.kt:
```diff
@@ -52,6 +55,36 @@ object Permissions {
+    private fun shouldRequestNotificationPermissions(context: Context): Boolean {
+        val permission = postNotification ?: return false // Null if below API 33
+        val grantedStatus = ContextCompat.checkSelfPermission(context, permission)
+        return (grantedStatus != PackageManager.PERMISSION_GRANTED)
```
This was a good point: I've implemented a variation of your suggestion after discovering a case where your proposed logic fails. See the long write-up I have below for more information. Thanks!

ericli3690 commented:
(First, a small sidenote: when I say that the user has "conclusively" rejected a permission, that means on API <30 that they have clicked "don't ask again" on the system UI dialog, whereas on API 30+, that means that they have clicked "deny" on the system UI dialog *twice*.)

Major changes to this PR! Here's an explanation of my thought process / how all the changes came to be:

Previously, when the user expressed a desire to grant notification permissions, we would eagerly launch the system UI permission request dialog. However, if the user had already conclusively rejected the permission, this would fail. To handle this, I added a callback to the system UI dialog launcher to check if the permission was actually granted. If it wasn't, then I would open the OS settings menu and ask the user to manually grant the permission. The assumption I made was that no reasonable user would click to toggle a reminder on (in the AnkiDroid UI's PermissionsBottomSheet) and then immediately press "deny" afterwards.

However, as Mike has pointed out, this has multiple flaws. First, my assumption above might be wrong: some users might legitimately click on the BottomSheet toggle, then change their minds after pressing the system UI text and click "deny". Second, as Mike says, if we add more places in the app where the BottomSheet can be launched from, we'll start to hit scenarios where the user encounters all three launch points and clicks "deny" at all three. The first two rejections will cause the rejection to be conclusive, and the third one will need to fall back to the OS settings launch. And, in general, we should not be pestering the user with three permission request BottomSheets if they've already denied it twice.

The fact that the OS settings launches is not necessarily bad: I would argue that if the user clicks on the toggle for the third occurrence of a notifications permission BottomSheet, they probably know what they're doing and have perhaps changed their minds about granting notification permissions, so we should ignore Android telling us they've "conclusively" rejected the permission and offer them the OS settings menu. The issue is performing this OS settings screen launch *after* first trying to launch the system UI permissions request dialog. As Mike says, launching that dialog when Android views the user as having conclusively denied the permission is likely stored in Play Console statistics and may decrease Play Store discoverability.

The obvious solution is to check if the user has conclusively denied the permission *before* trying to launch the dialog, rather than relying on the fallback callback after the system UI dialog attempts to open. However, this requires having some way to determine conclusive rejection, which -- as Mike as noted -- is impossible without internal Pref state management given the limited power of `shouldShowRequestPermissionRationale`. `#ThanksAndroid`

Fair enough. Well, since I've already added `reminderNotifsRequestShown` and `syncNotifsRequestShown` Prefs in this PR to check if the notification request BottomSheet has been shown, we can just check if at least one of them is set to `true`, as Mike suggested, right?

Not so fast. Those flags check if the *BottomSheet* has been launched, not if the *system UI dialog* has been triggered. A user might trigger the *BottomSheet* and then immediately dismiss it without clicking on the system UI dialog at all. That would cause Mike's suggested code to believe that permissions have been conclusively rejected, as `shouldShowRequestPermissionRationale` will still return false since the permission has never actually been "requested" yet (due to the system UI dialog not launching).

Hence we need a new Prefs flag to check if the system UI dialog has been launched for notification permission requesting. With this flag now in place, we are able to distinguish State 2 and State 4, which Mike identified [here](https://github.com/ankidroid/Anki-Android/pull/19167#discussion_r2524540116).

However, while implementing this change, I noticed that the audio recording permission is also very similar to the notifications permission. Via AllPermissionsExplanationFragment, the user is able to repeatedly click and deny it, also potentially leading to a decrease in Play Store discoverability. Hence, I also implemented a new Prefs flag to track if the system UI permissions request dialog has been launched for the audio recording permission and refactored AllPermissionsExplanationFragment.

Having done that, it became apparent that I had substantial duplicated code. I've decided to abstract out my work so that future developers can implement further permission requests similarly; hence the creation of `isUserOpenToPermission` and `requestPermissionThroughDialogOrSettings`. With that all in place, the permissions request launcher is now never triggered in any scenario if the user has already conclusively denied the permission; instead, the OS settings menu is immediately launched.

This also then nicely allowed me to implement Mike's suggested warning dialog that persistently shows up in the ScheduleReminders fragment if notification permissions are not granted. I've decided to implement it using a Snackbar, as that feels like the simplest out-of-the-box solution. An "enable" button on the Snackbar, when pressed, calls `requestPermissionThroughDialogOrSettings`.

A final wrinkle: helper methods for launching the system UI permission request dialog used to be protected within PermissionsFragment, but since ScheduleReminders now needs to use them, I've moved them to Permissions and made them public.

Okay! I think that explains all the changes here. Ready for review!

ericli3690 commented:
Thanks Mike for the quick review and really kind words! Cheers 😁

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/account/LoginFragment.kt:
```diff
@@ -218,8 +212,6 @@ class LoginFragment : Fragment(R.layout.my_account) {
                             activity.setResult(RESULT_OK)
                             activity.finish()
                         } else {
-                            AccountActivity.checkNotificationPermission(requireContext(), notificationPermissionLauncher)
```
Let's get this in at the start of 2.24: this is a minor lapse in functionality

Just to confirm: notifications are required to perform a background sync of AnkiWeb media

ericli3690 replied:
Sorry, just to double check, there's nothing actionable on my end here, right? With my new code, new users should still see a notification permission request the first time they perform a sync, it'll just be done via `DeckPicker` rather than `LoginFragment`.

ericli3690 replied:
Hm, I understand what you mean now, and after thinking it through a bit more I now think that I was too hasty in moving the logic for checking notification permissions from the login pathway to the sync pathway. I've now decided to move it back. Every time the user logs into an account, we check if we should ask for notification permissions, and if so, then we display the permissions dialog. Previously, if the user logged in via the settings menu and not via the introduction activity / DeckPicker, then the permissions dialog would never show up, when really it probably should. With this fix, that's now rectified.

Specifically, I've confirmed the permissions bottom sheet shows up in three primary scenarios:
1. The user opens the app for the first time, presses "Get Started", presses sync, then logs in. A sync automatically starts and the permission dialog shows up upon a successful login. See the first video below.
2. The user opens the app for the first time, presses "Get Started", goes to settings, then logs in. The permission dialog shows up upon a successful login. No sync happens, as was the behaviour before. See the second video below.
3. The user opens the app for the first time, presses "Sync with AnkiWeb", and logs in. A sync automatically starts and the permission dialog shows up upon a successful login. This is the scenario you've outlined above. See the third video below.

Are there any other scenarios I'm missing that I should test?

As for whether or not this is where we *should* be showing a permissions dialog... well, I personally think it's not a bad idea. It's true that notifications are only required for a background media sync, but a background media sync happens in the background, which means we can't show any dialog to the user in the foreground when it initially happens. Best to ask for permissions the moment syncing becomes possible so we don't have to handle it later. At least, that's my opinion. Let me know if you have ideas for better places to put this code.

https://github.com/user-attachments/assets/f03ec83b-3aec-4437-9f36-a31b0a4dc5ff


https://github.com/user-attachments/assets/38cd8bf7-03e8-4ec5-aade-35d29a3bc789


https://github.com/user-attachments/assets/e30fa5ab-9e61-4491-9686-ebc1355eaa86

david-allison replied:
Video 2: might as well close the bottom sheet fragment once permissions are granted

ericli3690 replied:
Good point; I've modified the code so that if permissions are granted, the bottom sheet will auto-dismiss itself.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/utils/Permissions.kt:
```diff
@@ -52,6 +67,89 @@ object Permissions {
+     * has been requested before. Required because [ActivityCompat.shouldShowRequestPermissionRationale]
+     * only works correctly after the first request.
+     */
+    fun isUserOpenToPermission(
```
unsure on the name here

ericli3690 replied:
Good point. I've switched it up to `canPermissionBeRequested`, which is hopefully more straightforward. Let me know if you have any other suggestions!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/InitialActivity.kt:
```diff
@@ -213,6 +214,9 @@ enum class PermissionSet(
     APP_PRIVATE(emptyList(), null),
+
+    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
+    NOTIFICATIONS(Permissions.postNotification?.let { listOf(it) } ?: emptyList(), NotificationsPermissionFragment::class.java),
```
From what I can tell, yes, this `?.let` will only run every time the app process spins up and initializes the enum. It should be safe since the only place `PermissionSet.NOTIFICATIONS` is ever used is `Permissions.showNotificationsPermissionBottomSheetIfNeeded`, which gates the usage behind checks that the API version is above 33. If I understand API versions correctly, a device can't change its API version without a full OS update, which should reload the AnkiDroid process, meaning the permission string list held within `PermissionsSet.NOTIFICATIONS` should never become desynchronized from reality.

For reference, `Permissions.postNotification` is:

```kotlin
    val postNotification =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            Manifest.permission.POST_NOTIFICATIONS
        } else {
            null
        }
```

Just to be more explicit, though, I've now modified this to get rid of the null-checking logic. It now reads:

```kotlin
@RequiresApi(Build.VERSION_CODES.TIRAMISU)
NOTIFICATIONS(listOf(Manifest.permission.POST_NOTIFICATIONS), NotificationsPermissionFragment::class.java),
```

where the direct usage of `Manifest.permission.POST_NOTIFICATIONS` is valid due to our pre-existing conditioning on `@RequiresApi(Build.VERSION_CODES.TIRAMISU)`. Am I understanding your concern about a bug in this line correctly?

ericli3690 commented:
Addressed David's feedback:
- Implemented early returns in ScheduleReminders's `checkForNotificationPermissions` method
- Cleaned up some docstrings in Prefs
- Deleted unnecessary null checks in InitialActivity's PermissionSet declaration
- Renamed `isUserOpenToPermission` to `canPermissionBeRequested`

Created PermissionsTest to test `requestPermissionThroughDialogOrSettings` and `showNotificationsPermissionBottomSheetIfNeeded`. In particular, all four scenarios that Mike outlines [here](https://github.com/ankidroid/Anki-Android/pull/19167#discussion_r2524540116):

> 1- we have the permission (checkSelfPermission === true, shouldShowRequestPermissionRationale === false)
2- we do not have the permission and have not asked (checkSelfPermission === false, shouldShowRequestPermissionRationale === false)
3- we do not have the permission, have asked, and we can still request it (checkSelfPermission === false, shouldShowRequestPermissionRationale === true)
4- we do not have the permission and we can not request it (user selected "don't ask again", or simply denied twice in new android) (checkSelfPermission === false, shouldShowRequestPermissionRationale === false)

...are now tested. I also considered making an instrumented test to check that the dialog shows up as expected when the user syncs from the DeckPicker page, but decided against it as that only works if the user is logged into an account. I couldn't find a log-user-in test utility method, and writing one from scratch would probably not be worth the time investment for this one validation path. I might write an instrumented test to check for notification permission granting via the ScheduleReminders path, but if I do I'll add that in later on, together with the rest of the instrumented tests I have planned for that feature (ex. AddEditReminderDialog, reminder CRUD, etc.).

mikehardy commented:
> I found a really tricky bug in DeckPicker and have now fixed it by adding a new fix commit to the back of this PR.

Fairly epic, nice one

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanationFragment.kt:
```diff
@@ -80,16 +77,27 @@ class AllPermissionsExplanationFragment : PermissionsFragment(R.layout.all_permi
             notificationPermission.apply {
                 isVisible = true
-                offerToGrantOrRevokeOnClick(permissionRequestLauncher, arrayOf(it))
+                revokeIfGrantedOnClickElse {
```
the name is a little confusing - the 'happy path' is a user granting permissions on this screen. This method does both granting and revocation, but the method name focuses on the revocation

A comment above would suffice, but maybe the method name should change as well
